staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go

```diff
@@ -0,0 +1,169 @@
+	syncRequeueRateLimited
+	syncNothing
+)
+
+// OpenAPIAggregationManager is the interface between this controller and OpenAPI Aggregator service.
+type OpenAPIAggregationManager interface {
+	AddUpdateApiService(handler http.Handler, apiService *apiregistration.APIService) error
+	UpdateApiServiceSpec(apiServiceName string, spec *spec.Swagger, etag string) error
+	RemoveApiServiceSpec(apiServiceName string) error
+	GetApiServiceInfo(apiServiceName string) (handler http.Handler, etag string, exists bool)
+}
+
+// OpenAPIAggregationController periodically check for changes in OpenAPI specs of APIServices and update/remove
+// them if necessary.
+type OpenAPIAggregationController struct {
+	openAPIAggregationManager OpenAPIAggregationManager
+	queue                     workqueue.RateLimitingInterface
+	downloader                *openAPIDownloader
+
+	// To allow injection for testing.
+	syncHandler func(key string) (syncAction, error)
+}
+
+func NewOpenAPIAggregationController(downloader *openAPIDownloader, openAPIAggregationManager OpenAPIAggregationManager) *OpenAPIAggregationController {
+	c := &OpenAPIAggregationController{
+		openAPIAggregationManager: openAPIAggregationManager,
+		queue: workqueue.NewNamedRateLimitingQueue(
+			workqueue.NewItemExponentialFailureRateLimiter(successfulUpdateDelay, failedUpdateMaxExpDelay), "APIServiceOpenAPIAggregationControllerQueue1"),
+		downloader: downloader,
+	}
+
+	c.syncHandler = c.sync
+
+	return c
+}
+
+func (c *OpenAPIAggregationController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	glog.Infof("Starting OpenAPIAggregationController")
+	defer glog.Infof("Shutting down OpenAPIAggregationController")
+
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (c *OpenAPIAggregationController) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (c *OpenAPIAggregationController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	defer c.queue.Done(key)
+	if quit {
+		return false
+	}
+
+	action, err := c.syncHandler(key.(string))
+	if err == nil {
+		c.queue.Forget(key)
+	} else {
+		utilruntime.HandleError(fmt.Errorf("loading OpenAPI spec for %q failed with: %v", key, err))
+	}
+
+	switch action {
+	case syncRequeue:
+		c.queue.AddAfter(key, successfulUpdateDelay)
+	case syncRequeueRateLimited:
+		c.queue.AddRateLimited(key)
+	case syncNothing:
+	}
+
+	return true
+}
+
+func (c *OpenAPIAggregationController) sync(key string) (syncAction, error) {
+	handler, etag, exists := c.openAPIAggregationManager.GetApiServiceInfo(key)
+	if !exists || handler == nil {
+		return syncNothing, nil
+	}
+	returnSpec, newEtag, httpStatus, err := c.downloader.Download(handler, etag)
+	switch {
+	case err != nil:
+		return syncRequeueRateLimited, err
+	case httpStatus == http.StatusNotFound || returnSpec == nil:
```

Comment:
There's no case that handles `http.StatusNotModified`.

I believe this is why after some period of time we get the error message(s):

```
controller.go:107] loading OpenAPI spec for "v2alpha1.batch" failed with: OpenAPI spec does not exists
```
Presumably it works OK first time but any subsequent downloads actually return a 304 (i.e., StatusNotModified).

Reply:
Yep, that case should not have been removed in the refactoring. I added it back with a no-op.

Reply:
`v2alpha1.batch` should not be downloaded in the first place as a local service.

Reply:
Those are local specs. They have handler (kubeApiServer) so they will be downloaded. But it is fast and one time. This can be improved for sure.

Reply:
but the local specs have there special `localDelegateChainNamePattern` key, don't they? Will they appear twice in the map then?